src/utilities/data-formatter.ts

```diff
@@ -1,54 +1,65 @@
+
+/**TODO: might need to be extended to include string[][] */
+type Values = string[] | number[] | string;
+
 /**
  *
- * @param {String[]|Number[]} values
+ * @param {String|String[]|Number[]} values
  * @param maxThreshold
  * @constructor
  */
-class DataFormatter {
-  constructor(values, maxThreshold = 0) {
+export class DataFormatter {
+
+  indexTable: { [key: string]: number, [key: number]: number } = {};
+  characterTable: { [key: number]: string | number | null } = {};
+  characters: (string | number)[] = [];
+  specialIndexes: number[] = [];
+
+  constructor(private values?: Values | undefined, maxThreshold = 0) {
     if (values === undefined) return;
 
     this.values = values;
     // go over all characters and keep track of all unique ones seen
     // count up all characters
-    this.indexTable = {};
-    this.characterTable = {};
-    this.characters = [];
-    this.specialIndexes = [];
+
     this.buildCharactersFromIterable(values);
     this.buildTables(maxThreshold);
   }
 
-  buildCharactersFromIterable(values) {
-    const tempCharactersTable = {};
+  buildCharactersFromIterable(values: Values) {
+    const tempCharactersTable: any = {};
     for (
       let dataFormatterIndex = 0, dataFormatterLength = values.length;
       dataFormatterIndex < dataFormatterLength;
       dataFormatterIndex++
     ) {
       const characters = values[dataFormatterIndex];
 
-      if (characters.hasOwnProperty('length')) {
+      if ((characters as any).hasOwnProperty('length')) {
```

Comment:
This `any` is a bit concerning.

Reply:
"This any is a bit concerning."

We don't need the any, but I put it there as a warning that the type is not what we think it is.

Checking for the existence of the 'length' property does not help typescript narrow the type, so some casting is needed (next line after the check).

The reason why characters has the wrong type is that I typed 'Values' as 
`type Values = string[]  | number[] | string;`
But I confirmed that during runtime it can actually also be string[][], so:

`type Values = string[] | string[][] | number[] | string;`

But if I add this option it causes issues with this line:

`const character = values[dataFormatterIndex];`
and then
` this.characters.push(character);`

These lines looks a bit strange, since they are re-indexing into values instead of using the 'characters' variable which we now know is a number.

I did not want to change these lines in this commit (and before we get the tests passing again)

If you prefer I can add the nested array option to 'Values' , and then change the code so we are not re-indexing into values, it sounds like a good idea to me.